src/lib/optimizer/strategy/nested_expression_rule.cpp

```diff
@@ -0,0 +1,163 @@
+  auto column_expressions = node->column_expressions();
+  column_expressions.erase(column_expressions.begin() + column_id);
+
+  auto projection_node = std::make_shared<ProjectionNode>(column_expressions);
+  node->replace_with(projection_node);
+}
+
+std::optional<DataType> NestedExpressionRule::_get_type_of_expression(
+    const std::shared_ptr<LQPExpression>& expression) const {
+  if (expression->type() == ExpressionType::Literal) {
+    return data_type_from_all_type_variant(expression->value());
+  }
+
+  if (!expression->is_arithmetic_operator()) {
+    return std::nullopt;
+  }
+
+  const auto type_left = _get_type_of_expression(expression->left_child());
+  const auto type_right = _get_type_of_expression(expression->right_child());
+
+  if (type_left == DataType::Null) return type_right;
+  if (type_right == DataType::Null) return type_left;
+  if (type_left != type_right) return std::nullopt;
+
+  return type_left;
+}
+
+template <typename T>
+AllTypeVariant NestedExpressionRule::_evaluate_expression(boost::hana::basic_type<T> type,
+                                                          const std::shared_ptr<LQPExpression>& expression) const {
+  if (expression->type() == ExpressionType::Literal) {
+    return AllTypeVariant(boost::get<T>(expression->value()));
+  }
+
+  if (!expression->is_arithmetic_operator()) {
+    return NULL_VALUE;
+  }
+
+  const auto& arithmetic_operator_function = arithmetic_operator_function_from_expression<T>(expression->type());
+
+  auto value = AllTypeVariant{};
+
+  const auto& left_child = expression->left_child();
+  const auto& right_child = expression->right_child();
+  const auto left_is_literal = left_child->type() == ExpressionType::Literal;
+  const auto right_is_literal = right_child->type() == ExpressionType::Literal;
+
+  if ((left_is_literal && variant_is_null(left_child->value())) ||
+      (right_is_literal && variant_is_null(right_child->value()))) {
+    // one of the operands is a literal null - early out.
+    value = NULL_VALUE;
+
+  } else if (left_is_literal && right_is_literal) {
+    value = AllTypeVariant(
+        arithmetic_operator_function(boost::get<T>(left_child->value()), boost::get<T>(right_child->value())));
+
+  } else if (right_is_literal) {
+    auto left_value = _evaluate_expression(type, left_child);
+    value =
+        AllTypeVariant(arithmetic_operator_function(boost::get<T>(left_value), boost::get<T>(right_child->value())));
+
+  } else if (left_is_literal) {
+    auto right_value = _evaluate_expression(type, right_child);
+    value =
+        AllTypeVariant(arithmetic_operator_function(boost::get<T>(left_child->value()), boost::get<T>(right_value)));
+
+  } else {
+    auto left_value = _evaluate_expression(type, left_child);
+    auto right_value = _evaluate_expression(type, right_child);
+    value = AllTypeVariant(arithmetic_operator_function(boost::get<T>(left_value), boost::get<T>(right_value)));
+  }
```

Comment:
Is this whole block (checking if left/right are literals) necessary? If we simply do `arithmetic_operator_function(_evaluate_expression(left), _evaluate_expression(right)`, wouldn't we end up with the same thing?